framework/util/page_guard_manager.cpp

```diff
@@ -199,6 +209,377 @@ void PageGuardManager::InitializeSystemExceptionContext(void)
 #endif
 }
 
+#if defined(USE_USERFAULTFD)
+bool PageGuardManager::UffdInit()
+{
+    // open the userfault fd
+    uffd_fd_ = syscall(SYS_userfaultfd, UFFD_USER_MODE_ONLY | O_CLOEXEC | O_NONBLOCK);
+    if (uffd_fd_ == -1)
+    {
+        GFXRECON_LOG_DEBUG_ONCE("syscall/userfaultfd: %s", strerror(errno));
+        return false;
+    }
+
+    // enable for api version and check features
+    struct uffdio_api uffdio_api;
+    uffdio_api.api      = UFFD_API;
+    uffdio_api.features = UFFD_FEATURE_PAGEFAULT_FLAG_WP;
+    if (ioctl(uffd_fd_, UFFDIO_API, &uffdio_api) == -1)
+    {
+        GFXRECON_LOG_ERROR("ioctl/uffdio_api: %s", strerror(errno));
+
+        uffd_fd_ = -1;
+        return false;
+    }
+
+    if (uffdio_api.api != UFFD_API)
+    {
+        GFXRECON_LOG_ERROR("unsupported userfaultfd api");
+
+        uffd_fd_ = -1;
+        return false;
+    }
+
+    const uint64_t required_features[] = { UFFD_FEATURE_PAGEFAULT_FLAG_WP };
+    for (size_t i = 0; i < sizeof(required_features) / sizeof(required_features[0]); ++i)
+    {
+        if ((uffdio_api.features & required_features[i]) != required_features[i])
+        {
+            GFXRECON_LOG_ERROR("unsupported userfaultfd feature: 0x%" PRIx64 "\n", required_features[i]);
+
+            uffd_fd_ = -1;
+            return false;
+        }
+    }
+
+    const uint64_t requested_ioctls[] = { 1 << _UFFDIO_REGISTER };
+    for (size_t i = 0; i < sizeof(requested_ioctls) / sizeof(requested_ioctls[0]); ++i)
+    {
+        if ((uffdio_api.ioctls & requested_ioctls[i]) != requested_ioctls[i])
+        {
+            GFXRECON_LOG_ERROR("unsupported userfaultfd ioctl: 0x%" PRIx64 "\n", requested_ioctls[i]);
+
+            uffd_fd_ = -1;
+            return false;
+        }
+    }
+
+    return true;
+}
+
+std::atomic<bool> PageGuardManager::uffd_stop_handler_thread = { false };
+
+bool PageGuardManager::UffdResetEntryProtection(MemoryInfo* memory_info)
+{
+    UffdUnregisterMemory(memory_info->shadow_memory, memory_info->shadow_range);
+
+    void* shadow_memory = memory_info->shadow_memory;
+    if (munmap(memory_info->shadow_memory, memory_info->shadow_range))
```

Comment:
I don't think you are supposed to unmap the memory first here. This introduces a race condition between the unmap and the new mmap.

Reply:
I wasn't aware of that and I am dealing with random segfaults that may be explained by this?
Any ideas?

Reply:
I think you can just remove the unmap call. Redoing the mmap call with MAP_FIXED should work and throw out the previously mapped in pages.

Reply:
Seems you're right, he random segfaults are gone!

However I am seeing random pages that although they trigger a pagefault event, when populating them `ioctl/uffdio_copy` triggers a `File exists` error. Indeed checking beforehand the pages with `mincore` these pages are already allocated.

I am not sure if these are responsible for the random visual artifacts (randomly corrupted textures and meshes) I am observing.
Also noting that I don't see neither the already allocated pages not the visual artifacts when trying this on a desktop game (recapturing a trace of a commercial title)

Reply:
Also on desktop the `munmap` + `mmap` combination also seems to work.
Perhaps there's a difference in the strategy the two OSes allocate pages